reactivesocket-cpp/src/Frame.h

```diff
@@ -100,13 +104,38 @@ std::ostream& operator<<(std::ostream&, const FrameHeader&);
 class FrameBufferAllocator {
  public:
   static std::unique_ptr<folly::IOBuf> allocate(size_t size);
+  static std::unique_ptr<folly::IOBufQueue> allocateQueue();
 
   virtual ~FrameBufferAllocator() = default;
 
  private:
   virtual std::unique_ptr<folly::IOBuf> allocateBuffer(size_t size);
 };
 
+class FrameMetadata {
+ public:
+  explicit FrameMetadata(std::unique_ptr<folly::IOBuf> metadataPayload)
+  : metadataPayload_(std::move(metadataPayload)) {}
+  /// Empty metadata
+  explicit FrameMetadata()
+  : metadataPayload_(folly::make_unique<folly::IOBuf>()) {}
+
+  void serializeInto(folly::io::QueueAppender& app);
+
+  /// if metadata is present, deserializes it into metadata
+  static bool deserializeFrom(folly::io::Cursor& cur,
+                              const FrameFlags& flags,
+                              folly::Optional<FrameMetadata>& metadata);
+  bool deserializeFrom(folly::io::Cursor& cur);
+
+  uint32_t size();
+
+  std::unique_ptr<folly::IOBuf> metadataPayload_;
+};
+std::ostream& operator<<(std::ostream&, const folly::Optional<FrameMetadata>&);
+
+static const folly::Optional<FrameMetadata> FrameMetadata_EMPTY { folly::none };
```

Comment:
FrameMetadata::kEmpty would have to be copied in the constructors, and because it has a unique_ptr member, copy construction is disabled by default. Does it mean I need to use a && reference or is there a simpler way to pass it around?